File: 203-remove-linked-list-elements/remove-linked-list-elements.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        # Create a dummy node that points to the head of the list
        dummy = ListNode(0)
        dummy.next = head
        # Initialize current pointer to the dummy node
        curr = dummy
        
        # Traverse the list until the end
        while curr.next:
            # If the next node's value matches the target value, skip it
            if curr.next.val == val:
                curr.next = curr.next.next
            else:
                # Otherwise, move the current pointer to the next node
                curr = curr.next
        
        # Return the new head of the list, which is dummy.next
        return dummy.next

File: 203-remove-linked-list-elements/remove-linked-list-elements.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} val
 * @return {ListNode}
 */
var removeElements = function(head, val) {
    let dummy = new ListNode(-1);
    dummy.next = head;
    let curr = dummy;

    while (curr && curr.next) {
        if (curr.next.val === val) {
            curr.next = curr.next.next; // skip the node with target value
        } else {
            curr = curr.next; // move forward
        }
    }

    return dummy.next;
};